### PR TITLE
WIP on combat...

### PR DESCRIPTION
**Pull request by miketwo**
Tried to make a few fixes. 

It would be better, imo, to get combat working end-to-end for a small number of cards/enemies before converting tons of things to classes.

### CODE REVIEW

File: entities.py [hunk 1]
```diff
@@ -124,11 +124,11 @@ def use_card(self, card, target: 'Enemy', exhaust, pile) -> None:
...
             return
         if card.target == 'Single':
-            card.apply(self, target)
+            card.apply(origin=self, target=target)
         elif card.target == 'Area':
-            card.apply(self, active_enemies)
+            card.apply(origin=self, enemies=active_enemies)
         elif card.target == 'Player':
-            card.apply(self)
+            card.apply(origin=self)
```
Comment (miketwo) [inline entities.py:131]: I suggest using keyword args in general, especially at the beginning when functions are changing often. If you rely just on the order of args, that can be more brittle.

File: test_combat.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    responses = iter("11\n11\n2\n")
+    with monkeypatch.context() as m:
+        m.setattr('builtins.input', lambda *a, **kw: next(responses))
+        m.setattr(helper, 'sleep', lambda _: None)
+        m.setattr(entities, 'sleep', lambda _: None)
+        m.setattr(items, 'sleep', lambda _: None)
+        helper.view.clear = replacement_clear_screen
+
+        # Run combat
+        combat.combat(game_map)
```
Comment (miketwo) [inline test_combat.py:45]: dedicated combat test. You can interact with it manually if you comment out the `builtins.input` line.

File: entities.py (hunk 1 above)
Comment (Vesper-arch) [reply entities.py:131]: Fair. I've come across problems where I change the order of a function's arguments and lots of things break because of it.

**Comment by Vesper-arch** [comment]
Currently saying goodbye to my mouse with Neovim. Wish me luck!

File: test_combat.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    acidslime = enemy_catalog.AcidSlimeS()
+    jawworm = enemy_catalog.JawWorm()
+
+    # Create combat object
+    combat = game.Combat(player, CombatTier.NORMAL, [acidslime, jawworm])
+
+    game_map = Mock()
+
+    # Patch the input
+    responses = iter("11\n11\n2\n")
```
Comment (Vesper-arch) [inline test_combat.py:36]: What's this?

File: test_combat.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+from unittest.mock import Mock
```
Comment (Vesper-arch) [inline test_combat.py:1]: What is this class supposed to do?

File: test_combat.py (hunk 3 above)
Comment (miketwo) [reply test_combat.py:36]: [iter is a built-in function that returns an iterable](https://www.programiz.com/python-programming/methods/built-in/iter) (like a generator expression). It's essentially the same as:

```
for c in "11\n11\n2\n"
  yield c
```

This allows it to be called repeatedly. We patch `input()` with it, so it gets called various times and we return a single character. So "11\n" is "play the first card, select the first target, press enter to continue". We do that twice. And then "2\n" is "play the 2nd card", which is a Defend card. This allows you to script the game inputs.

File: test_combat.py (hunk 4 above)
Comment (miketwo) [reply test_combat.py:1]: [I made a video on this when I was younger](https://www.youtube.com/watch?v=smPbDqGjFAI).

**Comment by miketwo** [comment]
Might want to rebase away the node commit (squash those 2 commits together) so that it doesn't increase the size of your repo when you merge.

**Comment by Vesper-arch** [comment]
what command do I use to do that?

**Comment by miketwo** [comment]
`git rebase -i message_bus_fixed` and then change the 4329bb2 commit to "s" for "squash". 

It's called interactive rebase, if you want to google it.

**Comment by miketwo** [comment]
If it works, there should be no change to the code, but `gitk --all` will show a shorter branch with the 432 commit removed. Then you can force push the branch to overwrite it.

**Comment by Vesper-arch** [comment]
uhhhhhhhhhhh. I think I did it? 
image

**Comment by Vesper-arch** [comment]
Well it seems to be updated on Github so I'll take that as a success.

**Comment by Vesper-arch** [comment]
Unless you have anything you want me to do on the pr, I'll merge it.

**Comment by miketwo** [comment]
lgtm

**Pull request merged**